src/core/QuoridorBoard.cs

```diff
@@ -8,26 +10,189 @@ namespace QuoridorEngine.Core
+            // Wall is vertical
+            else if (row % 2 == 0 && col % 2 != 0)
+                expRow = row - 2;
+
+            Debug.Assert(IsGridSquare(expRow, expCol));
+            return (expRow, expCol);
         }
 
-        public void AddWall(QuoridorPlayer player, int x, int y)
+        /// <summary>
+        /// Check if horizontal wall exists between two player squares 
+        /// </summary>
+        /// <param name="row">Player squares row</param>
+        /// <param name="col">Player square one column</param>
+        /// <param name="newCol">Player square two column</param>
+        /// <returns>True if wall exists</returns>
+        public bool CheckWallHorizontal(int row, int col, int newCol)
         {
-            usedWalls.Add((x, y));
-            player.DecreaseAvailableWalls();
+            Debug.Assert(col != newCol);
+            int dx = Math.Sign(newCol - col);
+            return usedWalls.Contains((row, col+dx));
         }
 
-        public void RemoveWall(QuoridorPlayer player, int x, int y)
+        /// <summary>
+        /// Check if vertical wall exists between two player squares 
+        /// </summary>
+        /// <param name="row">Player squares one row</param>
+        /// <param name="col">Player squares column</param>
+        /// <param name="newRow">Player square two row</param>
+        /// <returns>True if wall exists</returns>
+        public bool CheckWallVertical(int row, int col, int newRow)
         {
-            usedWalls.Remove((x, y));
-            player.IncreaseAvailableWalls();
+            Debug.Assert(row != newRow);
+            int dy = Math.Sign(newRow - row);
+            return usedWalls.Contains((row+dy, col));
+        }
+
+        /// <summary>
+        /// Check if provided wall forms a 
+        /// cross with an already existing one 
+        /// </summary>
+        /// <param name="row">First Wall Part Row</param>
+        /// <param name="col">First Wall Part Column</param>
+        /// <param name="expRow">Second Wall Part Row</param>
+        /// <param name="expCol">Second Wall Part Column</param>
+        /// <returns></returns>
+        public bool FormsWallCross(int row, int col, int expRow, int expCol)
+        {
+            Debug.Assert(IsGridSquare(row, col));
+            Debug.Assert(!IsPlayerSquare(row, col));
+            Debug.Assert(IsGridSquare(expRow, expCol));
+            Debug.Assert(!IsPlayerSquare(expRow, expCol));
+
+            return usedWalls.Contains((col, row)) && usedWalls.Contains((expCol, expRow));
+        }
+
+        /// <summary>
+        /// Add wall part to used walls
+        /// </summary>
+        /// <param name="row">Wall part row</param>
+        /// <param name="col">Wall part column</param>
+        public void AddWallPart(int row, int col)
+        {
+            Debug.Assert(IsGridSquare(row, col));
+            Debug.Assert(!IsPlayerSquare(row, col));
+
+            if (!usedWalls.Contains((row, col)))
```

Comment:
If in this case is unnecessary. First Assert the condition then do the operation without checking anything further.